### PR TITLE
Convert Shlwapi interop code to CSWin32

### PR DESCRIPTION
**Pull request by elachlan**
Relates to #7445

I am unsure on the `SHLoadIndirectString` changes and would love some feedback.

In MSBuild, we started with a smaller buffer and looped making it larger if the call failed. Also used `ArrayPool` in some spots. I guess it really depends on how large the returned information can technically get.

https://github.com/dotnet/msbuild/pull/7186

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7839)

### CODE REVIEW

**Comment by elachlan** [comment]
Usually most APIs have a length output which makes it possible to quickly ascertain the required length. But some don't and require using a loop and `ArrayPool`. If we know the max length in advance, this makes it a lot easier.

Example:
https://github.com/dotnet/msbuild/blob/cc2719b701ecf85687b69067bb0a9d9d55344cbe/src/Tasks/ComReference.cs#L407-L436

**Comment by elachlan** [comment]
@RussKie There is a failure based on the CI image `Unable to find directory for cmake 3.21.0;`

The test failures I am unsure if they are related or flaky tests?

**Review by RussKie — APPROVED** [review]
LGTM

File: src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.SHAutoComplete.cs [hunk 1]
```diff
@@ -2,23 +2,18 @@
...
     {
-        [DllImport(Libraries.Shlwapi, ExactSpelling = true)]
-        public static extern HRESULT SHAutoComplete(IntPtr hwndEdit, SHACF flags);
-
-        public static HRESULT SHAutoComplete(IHandle hwndEdit, SHACF flags)
+        public static HRESULT SHAutoComplete(IHandle<HWND> hwndEdit, SHELL_AUTOCOMPLETE_FLAGS flags)
         {
             HRESULT result = SHAutoComplete(hwndEdit.Handle, flags);
             GC.KeepAlive(hwndEdit);
             return result;
```
Comment (lonitra) [inline src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.SHAutoComplete.cs:13]: We want to use the following existing pattern: 

```suggestion
        public static HRESULT SHAutoComplete<T>(in T hwndEdit, SHELL_AUTOCOMPLETE_FLAGS flags) where T : IHandle<HWND>
        {
            HRESULT result = SHAutoComplete(hwndEdit.Handle, flags);
            GC.KeepAlive(hwndEdit.Wrapper);
            return result;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs [hunk 2]
```diff
@@ -283,11 +282,16 @@ Language ID
...
             {
-                var sb = new StringBuilder(512);
-                HRESULT res = Shlwapi.SHLoadIndirectString(layoutDisplayName, sb, (uint)sb.Capacity, IntPtr.Zero);
-                if (res == HRESULT.S_OK)
+                unsafe
                 {
-                    return sb.ToString();
+                    int len = 512;
+                    var ppvReserved = (void*)IntPtr.Zero;
+                    PWSTR buffer = stackalloc char[len];
```
[the remark this answers is not in the document]
Comment (elachlan) [reply src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs:289]: Would you like me to change it or is the current configuration okay?

[the remark this answers is not in the document]
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs:289]: Yes, please.

File: src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.SHAutoComplete.cs [hunk 3]
```diff
@@ -2,23 +2,18 @@
...
+    internal static partial class PInvoke
     {
-        [DllImport(Libraries.Shlwapi, ExactSpelling = true)]
-        public static extern HRESULT SHAutoComplete(IntPtr hwndEdit, SHACF flags);
-
-        public static HRESULT SHAutoComplete(IHandle hwndEdit, SHACF flags)
+        public static HRESULT SHAutoComplete<T>(in T hwndEdit, SHELL_AUTOCOMPLETE_FLAGS flags) where T : IHandle<HWND>
         {
             HRESULT result = SHAutoComplete(hwndEdit.Handle, flags);
             GC.KeepAlive(hwndEdit);
```
Comment (lonitra) [inline src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.SHAutoComplete.cs:12]:
```suggestion
            GC.KeepAlive(hwndEdit.Wrapper);
```

File: src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.SHAutoComplete.cs [hunk 4]
```diff
@@ -2,23 +2,18 @@
...
             return result;
         }
 
-        public static HRESULT SHAutoComplete(HandleRef hwndEdit, SHACF flags)
+        public static HRESULT SHAutoComplete(HandleRef<HWND> hwndEdit, SHELL_AUTOCOMPLETE_FLAGS flags)
```
Comment (lonitra) [inline src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.SHAutoComplete.cs:16]: If I'm remembering correctly, `HandleRef<HWND>` implements `IHandle<HWND>` so can this method actually be removed and we use just the above method?

Comment (elachlan) [reply src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.SHAutoComplete.cs:16]: Cool!

File: src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (elachlan) [reply src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs:289]: done.

File: src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs [hunk 5]
```diff
@@ -283,11 +282,18 @@ Language ID
...
             if (layoutDisplayName is not null)
             {
-                var sb = new StringBuilder(512);
-                HRESULT res = Shlwapi.SHLoadIndirectString(layoutDisplayName, sb, (uint)sb.Capacity, IntPtr.Zero);
-                if (res == HRESULT.S_OK)
+                unsafe
                 {
-                    return sb.ToString();
+                    var ppvReserved = (void*)IntPtr.Zero;
+                    Span<char> buffer = stackalloc char[512];
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs:288]: This is a better change than what I was suggesting earlier, thank you

**Review by lonitra — APPROVED** [review]
Looks good!

File: src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs (hunk 5 above)
Comment (elachlan) [reply src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs:288]: I was looking at the theming PR and noticed a similar scenario using this structure, so used it here.

**Pull request merged**